### PR TITLE
Downloading s3 and gcs files for now

### PR DESCRIPTION
**Pull request by AdityaHegde**
There is some issue with our integration with duckdb where successive imports directly from s3/gcs leads to an invalid state. No plugins work in this state, hence successive imports do not work.

For now downloading the files locally and importing them as files into duckdb. No issues here.

### CODE REVIEW

File: runtime/connectors/gcs/gcs.go [hunk 1]
```diff
@@ -70,3 +78,44 @@ type connector struct{}
...
+func getGcsUrlParts(path string) (string, string, string, error) {
+	u, err := url.Parse(path)
+	if err != nil {
+		return "", "", "", err
+	}
+
+	p := strings.Split(u.Path, "/")
+
+	return u.Host, strings.Replace(u.Path, "/", "", 1), p[len(p)-1], nil
+}
```
Comment (begelundmuller) [inline runtime/connectors/gcs/gcs.go:121]: It returns `u.Host` here but in `Consume` it's used as the bucket name?

File: runtime/connectors/gcs/gcs.go [hunk 2]
```diff
@@ -70,3 +78,44 @@ type connector struct{}
 func (c connector) Spec() connectors.Spec {
 	return spec
 }
+
+// Consume will eventually be added to the interface
+func Consume(ctx context.Context, source *connectors.Source) (string, error) {
+	conf, _ := ParseConfig(source.Properties)
```
Comment (begelundmuller) [inline runtime/connectors/gcs/gcs.go:84]: Probably handle and return the config error – maybe we add validation there in future

File: runtime/connectors/s3/s3.go [hunk 3]
```diff
@@ -76,3 +87,66 @@ type connector struct{}
 func (c connector) Spec() connectors.Spec {
 	return spec
 }
+
+// Consume will eventually be added to the interface
+func Consume(ctx context.Context, source *connectors.Source) (string, error) {
+	conf, _ := ParseConfig(source.Properties)
```
Comment (begelundmuller) [inline runtime/connectors/s3/s3.go:93]: Probably handle and return the config error – maybe we add validation there in future

File: runtime/connectors/s3/s3.go [hunk 4]
```diff
@@ -76,3 +87,66 @@ type connector struct{}
...
+
+	// The session the S3 Downloader will use
+	sess := session.Must(session.NewSession(&aws.Config{
```
Comment (begelundmuller) [inline runtime/connectors/s3/s3.go:96]: `Must` will panic if an error is returned ([see here](https://github.com/aws/aws-sdk-go/blob/d5e20b80023f98afe5a9200ffe5d026a020fa6ca/aws/session/session.go#L379)). Just handle and return the error from `session.NewSession` instead)

File: runtime/connectors/s3/s3.go [hunk 5]
```diff
@@ -76,3 +87,66 @@ type connector struct{}
...
+		Credentials: getAwsCredentials(conf),
+	}))
+
+	// Create a downloader with the session and default options
+	downloader := s3manager.NewDownloader(sess)
+
+	bucket, key, filename, err := getAwsUrlParts(conf.Path)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse path %s, %v", conf.Path, err)
+	}
```
Comment (begelundmuller) [inline runtime/connectors/s3/s3.go:107]: Like in GCS, what if bucket is set in the `conf`? Should we disable that option?

File: runtime/connectors/s3/s3.go [hunk 6]
```diff
@@ -76,3 +87,66 @@ type connector struct{}
...
+
+	// Write the contents of S3 Object to the file
+	n, err := downloader.Download(f, &s3.GetObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		return "", fmt.Errorf("failed to download file, %v", err)
+	}
+	fmt.Printf("file downloaded, %d bytes\n", n)
```
Comment (begelundmuller) [inline runtime/connectors/s3/s3.go:124]: Remove before merge

File: runtime/connectors/s3/s3.go [hunk 7]
```diff
@@ -76,3 +87,66 @@ type connector struct{}
...
+		return credentials.NewStaticCredentialsFromCreds(credentials.Value{
+			SessionToken: conf.AWSSession,
+		})
+	} else if conf.AWSKey != "" && conf.AWSSecret != "" {
+		return credentials.NewStaticCredentialsFromCreds(credentials.Value{
+			AccessKeyID:     conf.AWSKey,
+			SecretAccessKey: conf.AWSSecret,
+		})
+	}
+	return nil
```
Comment (begelundmuller) [inline runtime/connectors/s3/s3.go:140]: I assume returning nil will cause it to use system default credentials? (From `~/.aws` or something?)

**Review by begelundmuller — CHANGES_REQUESTED** [review]
This is actually pretty good, it gives us more power for dealing with ingestion than the pure DuckDB approach. We might never want to move back to DuckDB for S3/GCS.

In addition to the comments inline, a few quick todos to make this better:

- Use [`ioutil.TempFile`](https://yourbasic.org/golang/temporary-file-directory/) to get a filename to write the temporary files
- Make sure to delete the temporary files after they've been consumed
  - How about renaming `Consume` to `ConsumeAsFile(ctx, source, callback func(filename string) error) error`. Then it can call the callback with the filename, and clean it up afterwards (there are other alternatives here, this just seems simplest for now)
  - Actually, is it more correct to call this `Produce`, since the driver is doing the consumption? Haha
- This is a small thing, but instead of handling the `gcs` and `s3` cases separately in the DuckDB driver, now that you're using the baseline implementation, you could just have it as a `default` case in the `switch` in `Ingest`

File: runtime/connectors/s3/s3.go (hunk 7 above)
Comment (AdityaHegde) [reply runtime/connectors/s3/s3.go:140]: Yes that is correct.

File: runtime/connectors/gcs/gcs.go [hunk 8]
```diff
@@ -70,3 +78,44 @@ type connector struct{}
...
+	f, err := os.Create(filename)
+	if err != nil {
+		return "", fmt.Errorf("os.Create: %v", err)
+	}
+
+	rc, err := client.Bucket(bucket).Object(object).NewReader(ctx)
+	if err != nil {
+		return "", fmt.Errorf("Object(%q).NewReader: %v", object, err)
+	}
+	defer rc.Close()
```
[the remark this answers is not in the document]
Comment (AdityaHegde) [reply runtime/connectors/gcs/gcs.go:103]: 1. Bucket cannot be set in the conf right now.
2. I dont think it makes sense to accept the credentials from conf as it stands now. The cli way of setting the credentials is preferred. This pulls it from the cli's config.
3. This doesnt seem to apply for gcs.

File: runtime/connectors/gcs/gcs.go (hunk 1 above)
Comment (AdityaHegde) [reply runtime/connectors/gcs/gcs.go:121]: Hostname ends up being the bucket in this case when using `gs://` format.

File: runtime/connectors/s3/s3.go (hunk 5 above)
Comment (AdityaHegde) [reply runtime/connectors/s3/s3.go:107]: Similarly the bucket is part of the url from `Path`.

File: runtime/connectors/file/file.go [hunk 9]
```diff
@@ -67,3 +69,7 @@ type connector struct{}
 func (c connector) Spec() connectors.Spec {
 	return spec
 }
+
+func (c connector) ConsumeAsFile(ctx context.Context, source *connectors.Source, callback func(filename string) error) error {
+	return errdefs.NotImplemented(nil)
```
Comment (begelundmuller) [inline runtime/connectors/file/file.go:74]: This adds a dependency on `"github.com/docker/docker/errdefs"`. Better to do `errors.New("not implemented")`.

File: runtime/connectors/gcs/gcs.go [hunk 10]
```diff
@@ -79,34 +79,46 @@ func (c connector) Spec() connectors.Spec {
...
+		return fmt.Errorf("storage.NewClient: %v", err)
 	}
 	defer client.Close()
 
-	bucket, object, filename, err := getGcsUrlParts(conf.Path)
+	bucket, object, extension, err := getGcsUrlParts(conf.Path)
 
-	f, err := os.Create(filename)
+	f, err := os.CreateTemp(
+		os.Getenv("TMPDIR"),
```
Comment (begelundmuller) [inline runtime/connectors/gcs/gcs.go:97]: Use [`os.TempDir()`](https://pkg.go.dev/os#TempDir) instead so it's portable

File: runtime/connectors/gcs/gcs.go [hunk 11]
```diff
@@ -79,34 +79,46 @@ func (c connector) Spec() connectors.Spec {
...
+		fmt.Sprintf("%s*.%s", source.Name, extension),
+	)
 	if err != nil {
-		return "", fmt.Errorf("os.Create: %v", err)
+		return fmt.Errorf("os.Create: %v", err)
 	}
```
Comment (begelundmuller) [inline runtime/connectors/gcs/gcs.go:102]: After the error check, add:

```go
defer func() {
  f.Close()
  os.RemoveAll(f.Name())
}()
```

(Notice the invoking `()` parens at the end, always trips me up. You can also make them separate `defer`s, but then beware that they're invoked in LIFO order, so need to be reversed.)

Then you guarantee that cleanup occurs even in all the error return cases below.

Also remember to then remove the `os.RemoveAll(f.Name())` at the end of the function

Comment (begelundmuller) [reply runtime/connectors/gcs/gcs.go:102]: Same applies for s3

File: runtime/connectors/gcs/gcs.go (hunk 10 above)
Comment (begelundmuller) [reply runtime/connectors/gcs/gcs.go:97]: Same applies for s3

File: runtime/drivers/duckdb/connectors.go [hunk 12]
```diff
@@ -21,14 +21,13 @@ func (c *connection) Ingest(ctx context.Context, source *connectors.Source) erro
 	switch source.Connector {
 	case "file":
 		return c.ingestFile(ctx, source)
-	case "s3":
-		return c.ingestS3(ctx, source)
-	case "gcs":
-		return c.ingestGCS(ctx, source)
+		// TODO: figure out the issue with duckdb and perhaps use this instead of downloading
```
Comment (begelundmuller) [inline runtime/drivers/duckdb/connectors.go:24]: Never know what the future brings, but if you want, I'm fine with permanently removing those functions – this approach seems cleaner to me.

**Review by begelundmuller — APPROVED** [review]
Few minor changes left, then ready to merge. Approving as I don't think it'll require another review

File: runtime/drivers/duckdb/connectors.go (hunk 12 above)
Comment (AdityaHegde) [reply runtime/drivers/duckdb/connectors.go:24]: Ya sounds good. The old implementation should always be available through this PR

File: runtime/connectors/gcs/gcs.go (hunk 1 above)
Comment (ericpgreen2) [reply runtime/connectors/gcs/gcs.go]: Yeah, the `gs://` URI is what we'll prompt users to enter

**Comment by hamilton** [comment]
Correct me if I'm wrong, but I believe we misdiagnosed the underlying issues for this, right? The issue was due to threadpooling, not duckdb.

**Pull request merged**